trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java

```diff
@@ -1353,6 +1430,84 @@ private Result handleAtomicExchangeN(final IDispatcher main, final IASTFunctionC
 				applyMemoryOrder(loc, atomicBuilder.build(), memoryOrder.getLrValue().getValue())).build();
 	}
 
+	// https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html#index-_005f_005fatomic_005fcompare_005fexchange_005fn
+	// https://en.cppreference.com/w/c/atomic/atomic_compare_exchange
+	//
+	// The implementation below generates Boogie code that roughly follows this schema, where success and ptr_val are
+	// auxiliary variables:
+	//
+	// @formatter:off
+	// (evaluate arguments)
+	// atomic {
+	//  if weak
+	//    havoc success
+	//  else
+	//    success := true
+	//  ptr_val := read(ptr)
+	//  if success && ptr_val == read(expected):
+	//    write(desired, ptr)
+	//  else
+	//    success := false
+	// }
+	// if !success
+	//   write(ptr_val, expected)
+	// return success
+	// @formatter:on
+	private Result handleAtomicCompareExchangeN(final IDispatcher main, final IASTFunctionCallExpression node,
+			final ILocation loc, final String name) {
+		final IASTInitializerClause[] arguments = node.getArguments();
+		checkArguments(loc, 6, name, arguments);
+		final ExpressionResult pointer = mExprResultTransformer.dispatchPointerLValue(main, loc, arguments[0]);
+		final ExpressionResult expectedResult = mExprResultTransformer.dispatchPointerLValue(main, loc, arguments[1]);
+		final ExpressionResult desiredResult =
+				mExprResultTransformer.transformDecaySwitch((ExpressionResult) main.dispatch(arguments[2]), loc, node);
+		final ExpressionResult weakResult = mExprResultTransformer
+				.transformSwitchRexIntToBool((ExpressionResult) main.dispatch(arguments[3]), loc, node);
+		final ExpressionResult successMemoryOrder =
+				mExprResultTransformer.transformDispatchSwitchRexBoolToInt(main, loc, arguments[4]);
+		final ExpressionResult failureMemoryOrder =
+				mExprResultTransformer.transformDispatchSwitchRexBoolToInt(main, loc, arguments[5]);
+		final var resultBuilder = new ExpressionResultBuilder().addAllExceptLrValue(pointer, expectedResult,
+				desiredResult, weakResult, successMemoryOrder, failureMemoryOrder);
+
+		final var boolType = new CPrimitive(CPrimitives.BOOL);
+		final var success = mAuxVarInfoBuilder.constructAuxVarInfo(loc, boolType, AUXVAR.RETURNED);
+		final var pointerRead = mExprResultTransformer.readPointerValue(loc, pointer.getLrValue());
+		final var expectedRead = mExprResultTransformer.readPointerValue(loc, expectedResult.getLrValue());
+		final var pointerWrite = mExprResultTransformer.makePointerAssignment(loc, pointer.getLrValue(),
+				desiredResult.getLrValue().getValue());
+
+		final var successBoolean =
+				mExpressionTranslation.constructBinaryEqualityExpression(loc, IASTBinaryExpression.op_notequals,
+						success.getExp(), boolType, mExpressionTranslation.constructZero(loc, boolType), boolType);
+		final var atomicBody = new ExpressionResultBuilder().addAuxVarWithDeclaration(success)

```

Comment:
This is a huge single statement. I think, it should improve readability if you refactor it in smaller statements.

Reply:
I found it sometimes difficult to read other translation code if it first assembles lots of little pieces of Boogie (without context what they are useful for) and then assembles them at the end in some order.

So I tried to stick as closely as possible to the structure of the generated Boogie code here... though I can't say the result is necessarily much better :face_with_diagonal_mouth:

Reply:
Okay, I see. I think it looks better now. Maybe can improve the readability further to move the constructor call (maybe with the aux var) and the `build()` to separate statements. I think the indentation could be reduced in that case.